### PR TITLE
Dashboard Menu reducer refactor

### LINKED ISSUE #143: Refactor dashboard to a reducer structure.
The dashboard file is growing more and more and as well as its number of states, it is making sense to start handling these states with reducers.

https://reactjs.org/docs/hooks-reference.html#usereducer
https://kentcdodds.com/blog/the-state-reducer-pattern-with-react-hooks

### PR DESCRIPTION
**Pull request by matiasbavera**
## What's new
~~pre-requisite https://github.com/osrf/rmf-web/pull/180~~
fixes #143 
Replace https://github.com/osrf/romi-dashboard/pull/156

The dashboard file is growing more and more and as well as its number of states, it is making sense to start handling these states with reducers.

https://reactjs.org/docs/hooks-reference.html#usereducer
https://kentcdodds.com/blog/the-state-reducer-pattern-with-react-hooks

* Added main-menu reducer and actions.
* Wrap Reducer into a `customHook` to expose setters and getters following a "React" way of doing things.
* Implemented reducer on the dashboard component
* Refactor appbar
* Added dashboard-drawer
* Added reducer tests
* Update Hotkey tests with RTL
* Updated appbar tests with RTL
* Added tests for dashboard-drawer

* Improved Dashboard render performance

new
image

old
image

## Self-checks

- [X] I'm familiar with and follow this [ Typescript guideline](https://basarat.gitbook.io/typescript/styleguide)
- [x] I added unit-tests for new components
- [x] I tried testing edge cases
- [ ] I tested the behavior of the components that interact with the backend, with an e2e test

### CODE REVIEW

File: packages/dashboard/package.json [hunk 1]
```diff
@@ -80,12 +83,14 @@
...
   },
   "devDependencies": {
+    "@testing-library/react-hooks": "^3.4.2",
     "@types/jest": "^26.0.13",
     "@types/react-dom": "^16.9.8",
     "api-server": "file:../api-server",
     "concurrently": "^5.3.0",
     "enzyme-to-json": "^3.5.0",
     "gh-pages": "^3.1.0",
+    "react-test-renderer": "^17.0.1",
```
Comment (matiasbavera) [inline packages/dashboard/package.json:93]: Add this as a dependency of `@testing-library/react-hooks`

File: packages/dashboard/package.json [hunk 2]
```diff
@@ -48,6 +48,9 @@
     "@msgpack/msgpack": "^2.1.0",
     "@osrf/romi-js-core-interfaces": "^0.1.4",
     "@testing-library/react": "^11.1.1",
+    "@testing-library/jest-dom": "^5.11.4",
+    "@testing-library/user-event": "^12.1.9",
+    "@testing-library/react-hooks": "^3.4.2",
```
Comment (koonpeng) [inline packages/dashboard/package.json:53]: didn't know about this library, it looks really useful!

File: packages/dashboard/src/components/tests/appbar.test.tsx [hunk 3]
```diff
@@ -1,117 +1,101 @@
...
   beforeEach(() => {
-    toggleShowOmniPanel = jest.fn();
-    showSettings = jest.fn();
-    showHelp = jest.fn();
+    const hookResult = renderHook(() => useMainMenuReducer(mainMenuInitialValues));
+    result = hookResult.result;
+    root = render(<AppBar reducerMainMenu={result.current} alarmState={null} />);
   });
 
+  // Should we keep the snapshot test?
```
Comment (koonpeng) [inline packages/dashboard/src/components/tests/appbar.test.tsx:29]: We can keep it if you find it useful.

File: packages/dashboard/src/components/tests/appbar.test.tsx [hunk 4]
```diff
@@ -1,117 +1,101 @@
...
+          <AppBar reducerMainMenu={result.current} alarmState={null} />
         </UserContext.Provider>
       </AuthenticatorContext.Provider>,
     );
-    root.find('button#user-btn').simulate('click');
-    root.find('li#logout-btn').simulate('click');
+    userEvent.click(root.getByLabelText('user-btn'));
+    userEvent.click(root.getByText('Logout'));
+    // root.find('button#user-btn').simulate('click');
+    // root.find('li#logout-btn').simulate('click');
```
Comment (koonpeng) [inline packages/dashboard/src/components/tests/appbar.test.tsx:98]: leftovers?

File: packages/dashboard/src/components/reducers/main-menu-reducer.tsx [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+    case MainMenuActionType.TourState:
+      return { ...state, tourState: action.payload };
+    default:
+      console.error('Unexpected action');
+      return state;
+  }
+};
+
+export interface ReducerMainMenuProps extends MainMenuState {
+  setCurrentView: (payload: number) => void;
```
[the remark this answers is not in the document]
Comment (koonpeng) [reply packages/dashboard/src/components/reducers/main-menu-reducer.tsx:83]: In order to make to work nicely with omnipanel navigation, we need to make sure the view stack is synchronized see

```tsx
  const pushOmniPanelView = React.useCallback(
    (view: OmniPanelViewIndex) => {
      stackNavigator.push(view);
      setCurrentView(view);
    },
    [stackNavigator, setCurrentView],
  );
```

We could introduce a couple more actions, `pushView`, `popView`, `resetView` for example

**Review by koonpeng — CHANGES_REQUESTED** [review]
looking good so far!

File: packages/dashboard/package.json [hunk 6]
```diff
@@ -15,7 +15,7 @@
     "start:api": "api-server --port 50002 --trajectoryServerUrl ws://localhost:8006 --publicKey e2e/certs/keycloak.pub",
     "build": "react-scripts build",
     "build:storybook": "build-storybook -s public",
-    "test": "react-scripts test",
+    "test": "react-scripts test --env=jest-environment-jsdom-sixteen",
```
[the remark this answers is not in the document]
Comment (matiasbavera) [reply packages/dashboard/package.json:18]: > Added the `jest-environment-jsdom-sixteen` package and a test configuration for a `typescript - react-scripts - react testing library` bug explained here: https://stackoverflow.com/questions/61036156/react-typescript-testing-typeerror-mutationobserver-is-not-a-constructor

Sorry, I added this in the PR description, I'm realizing now that it was not the best place =)

[the remark this answers is not in the document]
Comment (matiasbavera) [reply packages/dashboard/package.json:18]: My node version is 12.20

[the remark this answers is not in the document]
Comment (koonpeng) [reply packages/dashboard/package.json:18]: Does it work if we upgrade/downgrade CRA or testing-library?

File: packages/dashboard/src/components/tests/appbar.test.tsx (hunk 4 above)
Comment (matiasbavera) [reply packages/dashboard/src/components/tests/appbar.test.tsx:98]: Yeah, thanks!

File: packages/dashboard/src/components/tests/appbar.test.tsx (hunk 3 above)
Comment (matiasbavera) [reply packages/dashboard/src/components/tests/appbar.test.tsx:29]: In this case, it is not. Removing it

File: packages/dashboard/package.json (hunk 6 above)
[the remark this answers is not in the document]
Comment (matiasbavera) [reply packages/dashboard/package.json:18]: I'll create docker with different versions to check this

**Comment by matiasbavera** [comment]
Rebased onto master

File: packages/dashboard/src/components/reducers/main-menu-reducer.tsx [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+    setSettings: (payload: Settings) =>
+      dispatch({ type: MainMenuActionType.Settings, payload: payload }),
+    setShowHelp: (payload: boolean) =>
+      dispatch({ type: MainMenuActionType.ShowHelp, payload: payload }),
+    setShowHotkeysDialog: (payload: boolean) =>
+      dispatch({ type: MainMenuActionType.ShowHotkeysDialog, payload: payload }),
+    setShowOmniPanel: (payload: boolean) =>
+      dispatch({ type: MainMenuActionType.ShowOmniPanel, payload: payload }),
+    setShowSettings: (payload: boolean) =>
+      dispatch({ type: MainMenuActionType.ShowSettings, payload: payload }),
```
[the remark this answers is not in the document]
Comment (matiasbavera) [reply packages/dashboard/src/components/reducers/main-menu-reducer.tsx:131]: Yeah, you're right, it's inferred.

File: packages/dashboard/package.json (hunk 6 above)
[the remark this answers is not in the document]
Comment (matiasbavera) [reply packages/dashboard/package.json:18]: CRA v3.4.3 -> v3.0.3
CRA 3.4.3 -> v4.0.1
RTL v11.0.4 -> v10.2.1
RTL v11.0.4  -> v11.2.2


Tests are working correctly upgrading and downgrading these versions. There is an error when running `npm run start` on the api-server when we upgrade  to CRA 4.0.1:

Re-exporting a type when the '--isolatedModules' flag is provided requires using 'export type'.  TS1205

```
     5 | import { WebSocketMiddleware } from './websocket-connect';
     6 | 
  >  7 | export { CustomLogger as Logger };
       |          ^
     8 | 
     9 | export interface RpcRequest<T = unknown> {
    10 |   version: string;
 
```

So I think we should upgrade versions of RTL and CRA on a FUP PR and fix that issue too.

[the remark this answers is not in the document]
Comment (matiasbavera) [reply packages/dashboard/package.json:18]: https://github.com/osrf/romi-dashboard/issues/178

[the remark this answers is not in the document]
Comment (koonpeng) [reply packages/dashboard/package.json:18]: thanks, if I got it right, we shouldn't need to explicitly set the jest env anymore after we perform the upgrade? If possible I would like to avoid this as that may bring maintaince problems later on as jest or react-scripts updates its environment.

File: packages/dashboard/src/components/reducers/main-menu-reducer.tsx [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+  | MainMenuActionFormat<'showHotkeysDialog', MainMenuState['showHotkeysDialog']>
+  | MainMenuActionFormat<'showOmniPanel', MainMenuState['showOmniPanel']>
+  | MainMenuActionFormat<'showSettings', MainMenuState['showSettings']>
+  | MainMenuActionFormat<'popView'>
+  | MainMenuActionFormat<'pushView', MainMenuState['currentView']>
+  | MainMenuActionFormat<'resetView'>
+  | MainMenuActionFormat<'toggleHelp'>
+  | Partial<MainMenuActionFormat<'toggleHotkeys'>>
+  | Partial<MainMenuActionFormat<'toggleOmnipanel'>>
+  | Partial<MainMenuActionFormat<'toggleSettings'>>
```
Comment (koonpeng) [inline packages/dashboard/src/components/reducers/main-menu-reducer.tsx:60]: I don't think we need the `Partial` here as there is no payload in these actions

File: packages/dashboard/package.json (hunk 6 above)
[the remark this answers is not in the document]
Comment (matiasbavera) [reply packages/dashboard/package.json]: Sure, I rebase this branch onto https://github.com/osrf/rmf-web/pull/180. Now we have the upgrade here and solved the problem.

File: packages/dashboard/src/components/reducers/main-menu-reducer.tsx (hunk 8 above)
Comment (matiasbavera) [reply packages/dashboard/src/components/reducers/main-menu-reducer.tsx:60]: It was a leftOver thanks!

**Comment by matiasbavera** [comment]
I needed to refactor the reducer to fix a huge performance issue =/. I fixed it but I need to update about 28 tests.

File: packages/dashboard/src/components/dashboard.tsx [hunk 9]
```diff
@@ -118,32 +131,35 @@ export default function Dashboard(_props: {}): React.ReactElement {
...
+    settings,
+    showOmniPanel,
+    stackNavigator,
+    showSettings,
+    showHelp,
+    showHotkeysDialog,
+  } = mainMenuState;
+
+  // eslint-disable-next-line react-hooks/exhaustive-deps
+  const mainMenuDispatch = React.useMemo(() => _mainMenuDispatch, []);
```
Comment (koonpeng) [inline packages/dashboard/src/components/dashboard.tsx:149]: It might mean bad design if we have to disable eslint checks. Maybe you can try

1. return the dispatch directly
2. move `useMemo` inside `useMainMenuReducer`. Add dispatch to the deps, or add the eslint ignore there (with a comment explaining it) so that callers don't have to do this.

**Review by koonpeng — CHANGES_REQUESTED** [review]

File: packages/dashboard/src/components/dashboard.tsx (hunk 9 above)
Comment (matiasbavera) [reply packages/dashboard/src/components/dashboard.tsx]: Thanks! Yeah, I move them to the custom hook `useMainMenuReducer` I managed to memoize the state and the dispatch there. Looks cleaner now.

**Review by koonpeng — APPROVED** [review]
LGTM
